### DISCUSSION TITLE
Export to nerfstudio

### DISCUSSION THREAD

**Discussion by zeeyang** [introlab/rtabmap#1231] [General]
Hi. I'm trying to convert db generated by rtab iOS app to [nerfstudio format](https://docs.nerf.studio/quickstart/data_conventions.html). I just want to confirm a few things around opt_poses in Admin table:
- there are exact same number of opt_poses as Data rows, ordered by Data.id
- opt_poses are in world space, right hand z-up coordinate system

**Comment by matlabbe** [comment]
We use ROS [REP103](https://www.ros.org/reps/rep-0103.html) coordinate system for `opt_poses`: +x forward, +y left, +z up. For image frame, it is (like opencv) +x- right, +y down, +z forward.

To get the actual pose of the image frame, you have to multiply the corresponding `local_transform` (in [Data](https://github.com/introlab/rtabmap/blob/8019a2dc49c7a845f5bba4cb94f01e3acebdb32a/corelib/src/resources/DatabaseSchema.sql.in#L35) table) of that frame. In recent versions, the `calibration` blob should be deserialized like [this](https://github.com/introlab/rtabmap/blob/8019a2dc49c7a845f5bba4cb94f01e3acebdb32a/corelib/src/CameraModel.cpp#L539-L607).

Note also that for iOS, depending of the rotation of the device, [we rotate the image](https://github.com/introlab/rtabmap/blob/8019a2dc49c7a845f5bba4cb94f01e3acebdb32a/app/android/jni/CameraMobile.cpp#L271-L357), but that rotation is included in `local_transform`. If you hold the phone always in portrait mode, there would be no extra rotation added.

For convenience, you can use `rtabmap-databaseViewer` (or Tools->Edit Database in [main app](https://github.com/introlab/rtabmap/releases)) and do File->Export poses, and select "Camera frame" instead of "Robot frame" to get all poses of the image frames. To export the poses from `opt_poses` table, open Graph View and select "Local optimized" poses in the comboxbox with "Global Iterative" by default.

cheers,
Mathieu

**Comment by zeeyang** [comment]
Thank you @matlabbe! I finally figured out I had to mirror along x-axis to convert from OpenCV to Nerfstudio space. If anyone else trying the same thing the answer is `opt_pose * local_transform * [[1,0,0],[0,-1,0],[0,0,-1]]` (or rotate using quat).

A quick followup question, what is the coordinate space of exported mesh/pointcloud ply? Is it ROS REP103?

**Reply by matlabbe** [reply]
Yes it is, +x forward, +y left, +z up.
